specification/under_development/profiles/Base_Profile.md

```diff
@@ -0,0 +1,484 @@
+primarily useful for execution on simulators, debug symbols are both easy to
+ignore and may be useful to generate helpful error messages for compilation
+failures that happen only late in the process. We hence see no reason to
+disallow them from occurring in the bitcode but will not detail their use any
+further. We defer to existing resources for more information about how to
+generate and use debug symbols.
+
+## Entry Point Definition
+
+The bitcode contains the definition of the LLVM function that should be invoked
+when the program is executed, referred to as entry point in the rest of this
+profile specification. The name of this function may be chosen freely, as long
+as it is a valid [global
+identifier](https://llvm.org/docs/LangRef.html#identifiers) by LLVM standard.
+
+The entry point may not take any parameters and must return an exit code in the
+form of a 64-bit integer. The exit code `0` must be used to indicate a
+successful execution of the program.
+
+### Attributes
+
+The following custom attributes must be attached to the entry point function:
+
+- An attribute named `"entry_point"` identifying the function as the starting
+  point of a quantum program
+- An attribute name `"qir_profile"` with the value `"base_profile"` identifying
+  the profile the entry point has been compiled for
+- An attribute name `"output_labels"` with an arbitrary string value that
+  identifies the schema used by a [compiler
+  frontend](https://en.wikipedia.org/wiki/Compiler#Front_end) that produced the
+  IR to label the recorded output
+- An attribute named `"required_qubits"` indicating the number of qubits used by
+  the entry point
+- An attribute named `"required_results"` indicating the maximal number of
+  measurement results that need to be stored while executing the entry point
+  function
+
+Optionally, additional attributes may be attached to the entry point. Any custom
+function attributes attached to the entry point will be reflected as metadata in
+the program output; this includes both mandatory and optional attributes but not
+parameter attributes or return value attributes. This in particular implies that
+the [labeling schema](#output-recording) used in the recorded output can be
+identified by looking at the metadata in the produced output. See the
+specification of the [output schemas](../output_schemas/) for more information
+about how metadata is represented in the output schema.
+
+Custom function attributes will show up as part of an [attribute
+group](https://releases.llvm.org/13.0.1/docs/LangRef.html#attrgrp) in the IR.
+Attribute groups are numbered such that they can be easily referenced by
+multiple function definitions or global variables. Consumers of Base Profile
+compliant programs must not rely on a particular numbering, but instead look for
+function to which an attribute with the name `"entry_point"` is attached to
+determine which one to invoke when the program is launched.
```

Comment:
BTW, what if this "entry_point" attribute is given to several functions? I'm not sure we mentioned this is forbidden.

Reply:
I would have thought that the entry point would be treated specially, like program parameters. A program is compiled, and then the user might want to have the option to specify any one of several entry points at run-time.  How would you envision this entry point attribute would be specified?  Must I do this prior to compilation?  That seems to limit flexibility.  Of course, there may be reasons or precedent in this LLVM space that suggests this attribute approach.  Not objecting, but I am just raising the question ... would it be better to have a dynamic entry point that is selectable on execution, not fixed during compile to one entry point?

Reply:
This is in interesting question. I expect that in the future there will be profiles that permit for multiple entry points. However, in the end, I believe it is more practical to require a compile time annotation to indicate the entry point(s). The reason is that there is a lot of compilation happening to getting the program simplified to match a profile (this one or others) - e.g. we aggressively constant fold (think of arrays that are commonly used in a program, even when they are not directly supported by a backend) and we inline/unroll loops as needed. With aggressive inlining for example, it is rather impractical to compile an entire library/lot of functions as if every one could be an entry point. We may also actually include program arguments in the compilation if needed/if that makes sense. So in reality, even if we pretended to choose the entry point at runtime, I believe it would still need to do significant jit compilation. The piece that launches/does this jit compilation might as well add the entry point attribute. I believe that piece/logic necessarily runs to get down to a profile in the first place.

Reply:
@LaurentAjdnik Regarding the entry_point attribute potentially being attached in several places, I added the following paragraph: 

> Both the `"entry_point"` attribute and the `"output_labels"` attribute can only
be attached to a function definition; they are invalid on a function that is
declared but not defined. For the Base Profile, this implies that they can occur
only in one place.

This means that the program should fail profile validation if the entry point attribute occurs in more than one place for the Base Profile.